File: go/nh_stream_test.go
package hermetix

// NH PLUG 스트림 — 가짜 웹소켓 서버로 구독·ACK·체결·호가·통보·재접속 흐름과 픽스처(문서 기반) 파싱을 검증한다.

import (
	"encoding/json"
	"strings"
	"testing"
	"time"
)

type nhMsg struct {
	Header struct {
		Token  string `json:"token"`
		TrType string `json:"tr_type"`
	} `json:"header"`
	Body struct {
		TrCd  string `json:"tr_cd"`
		TrKey string `json:"tr_key"`
	} `json:"body"`
}

func parseNhMsg(t *testing.T, raw string) nhMsg {
	t.Helper()
	var m nhMsg
	if err := json.Unmarshal([]byte(raw), &m); err != nil {
		t.Fatal(err)
	}
	return m
}

// parseFrame - 매번 새 jsonFrame 으로 파싱한다 (json.Unmarshal 은 기존 맵에 키를 합치므로 변수 재사용 금지).
func parseFrame(t *testing.T, raw string) jsonFrame {
	t.Helper()
	var frame jsonFrame
	if err := json.Unmarshal([]byte(raw), &frame); err != nil {
		t.Fatal(err)
	}
	return frame
}

func newNhTestStream(t *testing.T, marketCd string) (*NhMarketStream, chan *serverConn) {
	t.Helper()
	srv, connections := newWsServer(t)
	stream := newNhMarketStream(wsURL(srv), marketCd, "", func() (string, error) { return "TOKEN", nil })
	t.Cleanup(func() { stream.Close() })
	return stream, connections
}

func TestNhStreamSubscribeAndTick(t *testing.T) {
	fx := loadStreamFixture(t, "nh")
	stream, connections := newNhTestStream(t, "UNT")
	ticks := make(chan TradeTick, 16)
	stream.SubscribeTrades([]string{"KRX:005940"}, func(tick TradeTick) { ticks <- tick })
	stream.Connect()

	conn := waitConn(t, connections, 5*time.Second)
	msg := parseNhMsg(t, conn.take(t))
	if msg.Header.Token != "TOKEN" || msg.Header.TrType != "1" || msg.Body.TrCd != "mc" || msg.Body.TrKey != "005940" {
		t.Fatalf("subscribe = %+v", msg)
	}
	conn.send(t, `{"header":{"tr_type":"1","tr_cd":"mc","rsp_cd":"00000","rsp_msg":"정상처리되었습니다"},"body":{"tr_key":["005940"]}}`)
	conn.send(t, fx.Frames[0])
	tick := waitTick(t, ticks)
	if tick.Symbol != "KRX:005940" || tick.Price.String() != "31750" || tick.Quantity.String() != "13" {
		t.Fatalf("tick = %+v", tick)
	}
	if tick.CumulativeVolume == nil || *tick.CumulativeVolume != 837624 {
		t.Fatalf("cumulative = %v", tick.CumulativeVolume)
	}
	assertDecimalEqual(t, "change", tick.Change, "2500")
	assertDecimalEqual(t, "changeRate", tick.ChangeRate, "0.0855")
	assertDecimalEqual(t, "ask", tick.AskPrice, "31750")
	assertDecimalEqual(t, "bid", tick.BidPrice, "31700")
	if !stream.IsConnected() {
		t.Fatal("connected 여야 한다")
	}
}

func TestNhStreamKrxChannels(t *testing.T) {
	stream, connections := newNhTestStream(t, "KRX")
	stream.SubscribeTrades([]string{"005930"}, func(TradeTick) {})
	if err := stream.SubscribeOrderBook([]string{"005930"}, func(OrderBookTick) {}); err != nil {
		t.Fatal(err)
	}
	stream.Connect()
	conn := waitConn(t, connections, 5*time.Second)
	got := map[string]bool{}
	for i := 0; i < 2; i++ {
		m := parseNhMsg(t, conn.take(t))
		got[m.Body.TrCd] = true
	}
	if !got["oc"] || !got["ob"] {
		t.Fatalf("KRX 채널은 oc/ob 여야 한다: %v", got)
	}
}

func TestNhStreamOrderBook(t *testing.T) {
	fx := loadStreamFixture(t, "nh")
	stream, connections := newNhTestStream(t, "UNT")
	books := make(chan OrderBookTick, 4)
	if err := stream.SubscribeOrderBook([]string{"KRX:005940"}, func(b OrderBookTick) { books <- b }); err != nil {
		t.Fatal(err)
	}
	stream.Connect()
	conn := waitConn(t, connections, 5*time.Second)
	if m := parseNhMsg(t, conn.take(t)); m.Body.TrCd != "mb" {
		t.Fatalf("book subscribe = %+v", m)
	}
	conn.send(t, fx.OrderBook.Frames[0])
	var book OrderBookTick
	select {
	case book = <-books:
	case <-time.After(5 * time.Second):
		t.Fatal("호가가 오지 않음")
	}
	best, _ := book.BestAsk()
	bid, _ := book.BestBid()
	if book.Symbol != "KRX:005940" || len(book.Asks) != 10 || len(book.Bids) != 10 ||
		best.Price.String() != "31800" || best.Quantity.String() != "668" || bid.Price.String() != "31750" || bid.Quantity.String() != "2899" {
		t.Fatalf("book = %+v", book)
	}
	assertDecimalEqual(t, "totalAsk", book.TotalAskQuantity, "33938")
	assertDecimalEqual(t, "totalBid", book.TotalBidQuantity, "13132")
}

func TestNhStreamOrderEventsAndReconnect(t *testing.T) {
	fx := loadStreamFixture(t, "nh")
	stream, connections := newNhTestStream(t, "UNT")
	events := make(chan OrderEvent, 8)
	stream.SubscribeTrades([]string{"005940"}, func(TradeTick) {})
	if err := stream.SubscribeOrderEvents(func(e OrderEvent) { events <- e }); err != nil {
		t.Fatal(err)
	}
	stream.Connect()
	conn := waitConn(t, connections, 5*time.Second)
	registered := map[string]string{}
	for i := 0; i < 3; i++ {
		m := parseNhMsg(t, conn.take(t))
		registered[m.Body.TrCd] = m.Body.TrKey
	}
	if _, ok := registered["mc"]; !ok || registered["d2"] != "" || registered["d3"] != "" {
		t.Fatalf("registrations = %v", registered)
	}
	if v, ok := registered["d2"]; !ok || v != "" {
		t.Fatalf("d2 tr_key 는 빈 문자열이어야 한다: %v", registered)
	}

	for _, frame := range fx.OrderEvents.Frames {
		conn.send(t, frame)
	}
	take := func() OrderEvent {
		select {
		case e := <-events:
			return e
		case <-time.After(5 * time.Second):
			t.Fatal("이벤트가 오지 않음")
			return OrderEvent{}
		}
	}
	accepted := take()
	if accepted.Type != OrderAccepted || !accepted.OrderIDMatches("30") || accepted.Symbol != "005940" || accepted.Side == nil || *accepted.Side != Buy {
		t.Fatalf("accepted = %+v", accepted)
	}
	assertDecimalEqual(t, "quantity", accepted.Quantity, "10")
	assertDecimalEqual(t, "price", accepted.Price, "35550")
	filled := take()
	if filled.Type != OrderFilled {
		t.Fatalf("filled = %+v", filled)
	}
	assertDecimalEqual(t, "fill quantity", filled.Quantity, "5")

	// 서버가 끊으면 재접속해 세 등록을 다시 보낸다
	_ = conn.conn.Close(1000, "bye")
	second := waitConn(t, connections, 10*time.Second)
	again := map[string]bool{}
	for i := 0; i < 3; i++ {
		again[parseNhMsg(t, second.take(t)).Body.TrCd] = true
	}
	if !again["mc"] || !again["d2"] || !again["d3"] {
		t.Fatalf("resubscribe = %v", again)
	}
}

func TestNhParsersEdgeCases(t *testing.T) {
	today := fixtureToday
	numeric := `{"header":{"tr_cd":"mc","tr_key":"005930"},"body":{"code":"005930","time":"093012","sign":5,"change":300,"price":71500,"chrate":0.42,"offer":71500,"bid":71400,"movolume":15,"new_volume":1234567}}`
	ticks := ParseNhTrade(parseFrame(t, numeric), today)
	if len(ticks) != 1 || ticks[0].Price.String() != "71500" || ticks[0].Quantity.String() != "15" {
		t.Fatalf("numeric ticks = %+v", ticks)
	}
	assertDecimalEqual(t, "change (falling)", ticks[0].Change, "-300")
	assertDecimalEqual(t, "changeRate (falling)", ticks[0].ChangeRate, "-0.0042")
	if got := ticks[0].Timestamp.In(kst).Format("15:04:05"); got != "09:30:12" {
		t.Fatalf("time = %s", got)
	}

	// itemgb 가 2(파생)이면 무시, 계좌 필터 불일치도 무시
	d2 := `{"header":{"tr_cd":"d2"},"body":{"itemgb":"2","accountno":"A","orderno":"1","issuecd":"005930","slbygb":"2","concgty":"1","concprc":"1","conctime":"090000","ucgb":"0","rejgb":"0"}}`
	if len(ParseNhOrderEvents(parseFrame(t, d2), today, "")) != 0 {
		t.Fatal("itemgb=2 는 걸러져야 한다")
	}
	d2 = strings.Replace(d2, `"itemgb":"2"`, `"itemgb":"1"`, 1)
	if len(ParseNhOrderEvents(parseFrame(t, d2), today, "B")) != 0 {
		t.Fatal("계좌가 다르면 걸러져야 한다")
	}
	if events := ParseNhOrderEvents(parseFrame(t, d2), today, "A"); len(events) != 1 || events[0].Type != OrderFilled {
		t.Fatalf("filled = %+v", events)
	}
	cancel := strings.Replace(d2, `"ucgb":"0"`, `"ucgb":"2"`, 1)
	if events := ParseNhOrderEvents(parseFrame(t, cancel), today, ""); len(events) != 1 || events[0].Type != OrderCanceled {
		t.Fatalf("canceled = %+v", events)
	}
	reject := strings.Replace(d2, `"rejgb":"0"`, `"rejgb":"1"`, 1)
	if events := ParseNhOrderEvents(parseFrame(t, reject), today, ""); len(events) != 1 || events[0].Type != OrderRejected {
		t.Fatalf("rejected = %+v", events)
	}
	modifyAccept := `{"header":{"tr_cd":"d3"},"body":{"itemgb":"1","orderno":"0000000031","orgordno":"0000000030","issuecd":"005930","slbygb":"1","ordergty":"3","orderprc":"70000","order_time":"100000"}}`
	if events := ParseNhOrderEvents(parseFrame(t, modifyAccept), today, ""); len(events) != 1 || events[0].Type != OrderAccepted || events[0].OriginalOrderID != "0000000030" || *events[0].Side != Sell {
		t.Fatalf("modify accept = %+v", events)
	}
}

func TestNhStreamFixture(t *testing.T) {
	fx := loadStreamFixture(t, "nh")
	if fx.Measured {
		t.Fatal("nh 픽스처는 문서 기반(measured=false)이어야 한다")
	}
	var ticks []TradeTick
	for _, raw := range fx.Frames {
		ticks = append(ticks, ParseNhTrade(parseFrame(t, raw), fixtureToday)...)
	}
	assertStreamTicks(t, ticks, fx)

	var books []OrderBookTick
	for _, raw := range fx.OrderBook.Frames {
		books = append(books, ParseNhOrderBook(parseFrame(t, raw), fixtureToday)...)
	}
	assertBooks(t, books, fx.OrderBook)

	var events []OrderEvent
	for _, raw := range fx.OrderEvents.Frames {
		events = append(events, ParseNhOrderEvents(parseFrame(t, raw), fixtureToday, "")...)
	}
	assertOrderEvents(t, events, fx.OrderEvents)
}

File: js/tests/nhStream.test.ts
/** NH PLUG 실시간 스트림 — 로컬 ws 서버로 프로토콜을 재생한다 (Kotlin NhMarketStreamTest 대응, 문서 기반). */
import assert from "node:assert/strict";
import { test } from "node:test";
import { NhMarketStream, parseNhOrderBook, parseNhOrderEvents, parseNhTrade } from "../src/brokers/nhStream.js";
import type { OrderBookTick, OrderEvent, TradeTick } from "../src/models.js";
import { Queue, TODAY, assertBooks, assertEvents, assertTrades, sleep, streamFixture, withServer } from "./wsHarness.js";

const fx = streamFixture("nh");
const TRADE_FRAME = fx.frames[0];
const BOOK_FRAME = fx.orderBook.frames[0];
const [D3_FRAME, D2_FRAME] = fx.orderEvents.frames;
const opts = (url: string, marketCd = "UNT") => ({ wsUrl: url, token: async () => "TOKEN", marketCd, accountNo: "" });

test("nh: 통합 설정은 mc 로 토큰을 실어 구독하고 ACK 를 처리하며, 체결 프레임을 요청 표기로 전달한다", () =>
  withServer(async (url, connections) => {
    const ticks = new Queue<TradeTick>();
    const stream = new NhMarketStream(opts(url));
    try {
      stream.subscribeTrades(["KRX:005940"], (t) => ticks.put(t));
      stream.connect();
      const conn = await connections.take();
      const subscribe = JSON.parse(await conn.received.take());
      assert.equal(subscribe.header.token, "TOKEN");
      assert.equal(subscribe.header.tr_type, "1");
      assert.equal(subscribe.body.tr_cd, "mc");
      assert.equal(subscribe.body.tr_key, "005940");
      conn.send('{"header":{"tr_type":"1","tr_cd":"mc","rsp_cd":"00000","rsp_msg":"정상처리되었습니다"},"body":{"tr_key":["005940"]}}');
      conn.send(TRADE_FRAME);
      const tick = await ticks.take();
      assert.equal(tick.symbol, "KRX:005940");
      assert.ok(tick.price.eq(31750));
      assert.ok(tick.quantity.eq(13));
      assert.equal(tick.cumulativeVolume, 837624);
      assert.ok(tick.change!.eq(2500));
      assert.ok(tick.changeRate!.eq("0.0855"));
      assert.equal(stream.isConnected, true);
    } finally { stream.close(); }
  }));

test("nh: KRX 설정은 oc·ob 채널을 쓴다", () =>
  withServer(async (url, connections) => {
    const stream = new NhMarketStream(opts(url, "KRX"));
    try {
      stream.subscribeTrades(["005930"], () => {});
      stream.subscribeOrderBook(["005930"], () => {});
      stream.connect();
      const conn = await connections.take();
      const codes = [JSON.parse(await conn.received.take()).body.tr_cd, JSON.parse(await conn.received.take()).body.tr_cd].sort();
      assert.deepEqual(codes, ["ob", "oc"]);
    } finally { stream.close(); }
  }));

test("nh: 호가 mb 프레임을 요청 표기 심볼의 10단계 호가창으로 전달한다", () =>
  withServer(async (url, connections) => {
    const books = new Queue<OrderBookTick>();
    const stream = new NhMarketStream(opts(url));
    try {
      stream.subscribeOrderBook(["KRX:005940"], (b) => books.put(b));
      stream.connect();
      const conn = await connections.take();
      assert.equal(JSON.parse(await conn.received.take()).body.tr_cd, "mb");
      conn.send(BOOK_FRAME);
      const book = await books.take();
      assert.equal(book.symbol, "KRX:005940");
      assert.equal(book.asks.length, 10);
      assert.equal(book.bids.length, 10);
      assert.ok(book.asks[0].price.eq(31800));
      assert.ok(book.asks[0].quantity.eq(668));
      assert.ok(book.bids[0].price.eq(31750));
      assert.ok(book.bids[0].quantity.eq(2899));
      assert.ok(book.totalAskQuantity!.eq(33938));
      assert.ok(book.totalBidQuantity!.eq(13132));
    } finally { stream.close(); }
  }));

test("nh: 주문 통보는 d2·d3 를 빈 tr_key 로 등록하고 접수→체결 이벤트로 전달한다", () =>
  withServer(async (url, connections) => {
    const events = new Queue<OrderEvent>();
    const stream = new NhMarketStream(opts(url));
    try {
      stream.subscribeOrderEvents((e) => events.put(e));
      stream.connect();
      const conn = await connections.take();
      const regs = [JSON.parse(await conn.received.take()), JSON.parse(await conn.received.take())];
      assert.deepEqual(regs.map((r) => r.body.tr_cd).sort(), ["d2", "d3"]);
      assert.ok(regs.every((r) => r.body.tr_key === "" && r.header.tr_type === "1"));
      conn.send(D3_FRAME);
      conn.send(D2_FRAME);
      const accepted = await events.take();
      assert.equal(accepted.type, "ACCEPTED");
      assert.equal(accepted.orderId, "0000000030");
      assert.equal(accepted.symbol, "005940");
      assert.equal(accepted.side, "BUY");
      assert.ok(accepted.quantity!.eq(10));
      assert.ok(accepted.price!.eq(35550));
      const filled = await events.take();
      assert.equal(filled.type, "FILLED");
      assert.ok(filled.quantity!.eq(5));
      assert.ok(filled.price!.eq(35550));
    } finally { stream.close(); }
  }));

test("nh: 서버가 끊으면 재접속해 구독과 통보 등록을 다시 보낸다", () =>
  withServer(async (url, connections) => {
    const stream = new NhMarketStream(opts(url));
    try {
      stream.subscribeTrades(["005930"], () => {});
      stream.subscribeOrderEvents(() => {});
      stream.connect();
      const first = await connections.take();
      for (let i = 0; i < 3; i++) await first.received.take();
      first.socket.close(1000, "bye");
      const second = await connections.take(10_000);
      const codes: string[] = [];
      for (let i = 0; i < 3; i++) codes.push(JSON.parse(await second.received.take()).body.tr_cd);
      assert.deepEqual(codes.sort(), ["d2", "d3", "mc"]);
    } finally { stream.close(); }
  }));

test("nh: 연결된 뒤 추가 구독은 즉시 전송된다", () =>
  withServer(async (url, connections) => {
    const stream = new NhMarketStream(opts(url));
    try {
      stream.connect();
      const conn = await connections.take();
      await sleep(50);
      stream.subscribeTrades(["000660"], () => {});
      assert.equal(JSON.parse(await conn.received.take()).body.tr_key, "000660");
    } finally { stream.close(); }
  }));

test("nh 파서: 숫자 타입·HHMMSS 시각·하락 부호·itemgb 필터·계좌 필터", () => {
  const numeric = { header: { tr_cd: "oc", tr_key: "005930" }, body: { code: "005930", time: "093012", sign: "5", change: 300, price: 71500, chrate: 0.42, offer: 71500, bid: 71400, movolume: 15, new_volume: 1234567 } };
  const tick = parseNhTrade(numeric, TODAY)[0];
  assert.ok(tick.price.eq(71500));
  assert.ok(tick.quantity.eq(15));
  assert.ok(tick.change!.eq(-300));
  assert.ok(tick.changeRate!.eq("-0.0042"));
  assert.equal(tick.cumulativeVolume, 1234567);
  assert.equal(parseNhTrade({ header: { tr_cd: "oc" }, body: { code: "005930", time: "093012" } }, TODAY).length, 0); // 가격 없음

  const d2 = JSON.parse(D2_FRAME);
  d2.body.itemgb = "2"; // 국내파생 → 무시
  assert.equal(parseNhOrderEvents(d2, TODAY).length, 0);
  d2.body.itemgb = "1";
  assert.equal(parseNhOrderEvents(d2, TODAY, "OTHER").length, 0); // 다른 계좌
  assert.equal(parseNhOrderEvents(d2, TODAY, "ACCOUNT NUMBER").length, 1);
  d2.body.ucgb = "2";
  assert.equal(parseNhOrderEvents(d2, TODAY)[0].type, "CANCELED");
  d2.body.ucgb = "1";
  assert.equal(parseNhOrderEvents(d2, TODAY)[0].type, "MODIFIED");
  d2.body.rejgb = "1";
  assert.equal(parseNhOrderEvents(d2, TODAY)[0].type, "REJECTED");

  const d3 = JSON.parse(D3_FRAME);
  d3.body.orgordno = "0000000029";
  assert.equal(parseNhOrderEvents(d3, TODAY)[0].originalOrderId, "0000000029");
  assert.equal(parseNhOrderBook({ header: { tr_cd: "mb" }, body: { code: "005930", hotime: "13:55:11", offer: "0", bid: "31750", bidrem: "1" } }, TODAY)[0].asks.length, 0); // 0 호가 제외
});

test("nh 픽스처: stream 섹션(문서 기반)을 기대값대로 파싱한다", () => {
  assert.equal(fx.measured, false);
  assertTrades(fx.frames.flatMap((f) => parseNhTrade(JSON.parse(f), TODAY)), fx.expected);
  assert.equal(fx.orderBook.measured, false);
  assertBooks(fx.orderBook.frames.flatMap((f) => parseNhOrderBook(JSON.parse(f), TODAY)), fx.orderBook.expected);
  assert.equal(fx.orderEvents.measured, false);
  assertEvents(fx.orderEvents.frames.flatMap((f) => parseNhOrderEvents(JSON.parse(f), TODAY)), fx.orderEvents.expected);
});
